### ISSUE TITLE
hide Windows console

### ISSUE THREAD

**Issue by Mithgol** [sindresorhus/module-requests#81]
Node.js scripts on Windows, unless run as services (via [`node-windows2`](https://www.npmjs.com/package/node-windows2), for example), usually appear in a visible console window such as the following:

![(screenshot)](https://cloud.githubusercontent.com/assets/1088720/20487715/94970800-b015-11e6-84f5-5de29552f11a.gif)

It might be good to have a module that is able to hide that window (entirely or to the tray), especially if the main interaction happens elsewhere (in [libui](https://github.com/parro-it/libui-node/tree/master/docs) GUI, for example).

**Comment by sindresorhus** [comment]
Can you share some code that reproduces it?

**Comment by sindresorhus** [comment]
@Mithgol No, I mean a Node.js script that opens the console like you describe. I've never seen that before.

**Comment by Qix-** [comment]
Node.js needs to be linked with `/SUBSYSTEM:windows` instead of `/SUBSYSTEM:console` for this to happen.

**Comment by Mithgol** [comment]
@sindresorhus The above console wasn't actually opened by a script, I've just run `cmd` manually and then `node` in its window.

I admit that if I simply run `node.exe .` command in a directory (from some GUI file manager), usually a Node.js console window opens and it might not look exactly like cmd's (for example, it has a different icon, which is Node's, and also usually default font settings). Here's a (slightly outdated) webserver script for example:

![(screenshot)](https://cloud.githubusercontent.com/assets/1088720/20511885/e90c9252-b08c-11e6-90ba-1f88471a8460.gif)

**Comment by sindresorhus** [comment]
Sounds like you want http://www.ntwind.com/software/hstart.html

**Comment by Mithgol** [comment]
That “Hidden Start” actually starts a hidden console window instead of hiding an existing console window.

(Though that's a very similar feature and it may indeed cover a lot of use cases.)

**Comment by sindresorhus** [comment]
I'm aware, but it seems like it solves your needs even better than what you asked for, unless I'm missing something?

**Comment by Mithgol** [comment]
Perhaps it does, but only the “hide entirely” part. Is there a similar tool that hides to tray, I wonder.

**Comment by oledid** [comment]
One way to hide a window in Windows, is to know the window handle (hWnd) and use the windows api (user32.dll) to hide it.

Example C# code:
```csharp
[DllImport("user32.dll")]
private static extern bool ShowWindow(IntPtr hwnd, int nCmdShow);
```

```csharp
const int SW_HIDE = 0;
IntPtr hwnd = ?

ShowWindow(hwnd, SW_HIDE);
```

I used [edge](https://www.npmjs.com/package/edge) in node to hide spawned windows from a node process once, but it was unreliable since I could not find out how to get hWnd other than to get hWnd from the current active window right after spawning the process.

**Comment by oledid** [comment]
I think I managed to get this working. Will look at making it into a module tomorrow.

**Comment by oledid** [comment]
https://www.npmjs.com/package/windows-api-show-window

https://github.com/oledid/windows-api-show-window

**Comment by oledid** [comment]
Rewrote this to use ffi instead of edge.

**Comment by HRcat** [comment]
https://stackoverflow.com/questions/42222004/run-exe-in-background
